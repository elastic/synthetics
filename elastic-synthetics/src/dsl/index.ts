import { Journey, JourneyCallback, JourneyOptions } from './journey'
import { Step, StepCallback } from './step'

class State {
    journeys: Journey[]
    currentJourney?: Journey

    constructor() {
        this.reset();
    }

    public reset() {
        this.journeys = [];
        this.currentJourney = undefined;
    }
}

export const state = new State();

export const journey = (options: JourneyOptions | string, callback: JourneyCallback) => {
    if (typeof options === 'string') {
        options = {name: options, id: options}
    }
    const j = new Journey(options, callback);
    state.journeys.push(j)
    state.currentJourney = j;

    j.callback(); // load steps
}

export const step = (name: string, callback: StepCallback) => {
    const step = new Step(name, callback);
    state.currentJourney.steps.push(step)
}