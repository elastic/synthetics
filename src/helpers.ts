import { red, green, yellow, cyan } from 'kleur/colors';
import os from 'os';
import path from 'path';
import { performance } from 'perf_hooks';

export function noop() {}

export function indent(lines: string, tab = '   ') {
  return lines.replace(/^/gm, tab);
}

export const symbols = {
  warning: yellow('⚠'),
  skipped: cyan('-'),
  succeeded: green('✓'),
  failed: red('✖'),
};

export function formatError(error: Error) {
  if (!(error instanceof Error)) {
    return;
  }
  const { name, message, stack } = error;
  return { name, message, stack };
}

export function generateTempPath() {
  return path.join(os.tmpdir(), `synthetics-${process.hrtime().toString()}`);
}

/**
 * As per the timings used in the Network Events from
 * Chrome devtools protocol
 */
export function getMonotonicTime() {
  const hrTime = process.hrtime();
  return hrTime[0] * 1 + hrTime[1] / 1e9;
}

/**
 * Timestamp at which the current node process began.
 */
const processStart = performance.timeOrigin;
export function getTimestamp() {
  return (processStart + now()) * 1000;
}

/**
 * Relative current time from the start of the current node process
 */
export function now() {
  return performance.now();
}

/**
 * Execute all the callbacks in parallel using Promise.all
 */
export async function runParallel(callbacks) {
  const promises = callbacks.map(cb => cb());
  return await Promise.all(promises);
}
