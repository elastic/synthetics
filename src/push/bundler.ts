/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import path from 'path';
import { stat, unlink, readFile } from 'fs/promises';
import { createWriteStream } from 'fs';
import * as esbuild from 'esbuild';
import archiver from 'archiver';
import { commonOptions } from '../core/transform';
import { SyntheticsBundlePlugin } from './plugin';

const SIZE_LIMIT_KB = 800;

function relativeToCwd(entry: string) {
  return path.relative(process.cwd(), entry);
}

export class Bundler {
  moduleMap = new Map<string, string>();
  constructor() {}

  async prepare(absPath: string) {
    const options: esbuild.BuildOptions = {
      ...commonOptions(),
      ...{
        entryPoints: [absPath],
        bundle: true,
        write: false,
        sourcemap: 'inline',
        external: ['@elastic/synthetics'],
        plugins: [SyntheticsBundlePlugin()],
      },
    };
    const result = await esbuild.build(options);
    if (result.errors.length > 0) {
      throw result.errors;
    }
    this.moduleMap.set(absPath, result.outputFiles[0].text);
  }

  async zip(outputPath: string) {
    return new Promise((fulfill, reject) => {
      const output = createWriteStream(outputPath);
      const archive = archiver('zip', {
        zlib: { level: 9 },
      });
      archive.on('error', reject);
      output.on('close', fulfill);
      archive.pipe(output);
      for (const [path, content] of this.moduleMap.entries()) {
        const relativePath = relativeToCwd(path);
        // Date is fixed to Unix epoch so the file metadata is
        // not modified everytime when files are bundled
        archive.append(content, {
          name: relativePath,
          date: new Date('1970-01-01'),
        });
      }
      archive.finalize();
    });
  }

  async build(entry: string, output: string) {
    await this.prepare(entry);
    await this.zip(output);
    const data = await this.encode(output);
    await this.checkSize(output);
    await this.cleanup(output);
    return data;
  }

  async encode(outputPath: string) {
    return await readFile(outputPath, 'base64');
  }

  async checkSize(outputPath: string) {
    const { size } = await stat(outputPath);
    const sizeKb = size / 1024;
    if (sizeKb > SIZE_LIMIT_KB) {
      throw new Error(
        `You have monitors whose size exceeds the ${SIZE_LIMIT_KB}KB limit.`
      );
    }
  }

  async cleanup(outputPath: string) {
    this.moduleMap = new Map<string, string>();
    await unlink(outputPath);
  }
}
