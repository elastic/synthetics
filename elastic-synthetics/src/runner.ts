import { state, journey, step } from './dsl';
import * as playwright from 'playwright';
import { Command } from 'commander';
import { readFileSync } from 'fs';
import { createInterface as createReadlineInterface } from 'readline'


const browserType = 'chromium'

export const runJourneys = async (suiteParams: any) => {
    console.debug("Running with suite params", suiteParams);
    const browser = await playwright[browserType].launch({ headless: false });
    const context = await browser.newContext()
    const page = await context.newPage()
    console.log(`Found ${state.journeys.length} journeys`)
    for (let i = 0; i < state.journeys.length; i++) {
        const journey = state.currentJourney = state.journeys[i]
        console.log(`Journey: ${journey.options.name}`)
        for (let j = 0; j < journey.steps.length; j++) {
            const step = journey.steps[j]
            console.log(`Step: ${step.name}`)
            await step.callback(page, suiteParams, {context, browser})
        }
    }

    await browser.close()
}

export const run = async (aroundHook: (runSteps: (suiteParams: any) => void, environment: string, suiteParams: any) => void = (r) => { r({}) }) => {
    const program = new Command;
    let argSuiteParams = {};
    program
        .command("run [file]")
        .option('-s, --suite-params <jsonstring>')
        .option('-e, --environment <envname>', 'e.g. production', 'development')
        .option('-j, --json', 'output newline delimited JSON')
        .option('--stdin', 'read script file input from stdin')
        .description("Run the tests")
        .action(async (file, options) => {
            if (options.suiteParams) {
                const parsedSuiteParams = JSON.parse(options.suiteParams);
                console.log("SPARAMS", parsedSuiteParams)
                argSuiteParams = {...argSuiteParams, ...parsedSuiteParams};
            }

            const singleMode = options.stdin || file;
            if (singleMode) {
                state.reset()
                const source = options.stdin ? await readStdin() : readFileSync(file, {encoding: 'utf8'});
                console.log("Running single script...", source)
                loadInlineScript(source, argSuiteParams);
            }

            console.log("Running...")

            aroundHook((suiteParams) => runJourneys(suiteParams), options.environment, argSuiteParams)
        })

    program.parse()
}

const readStdin = async (): Promise<string> => {
    let source = "";
    const rl = createReadlineInterface({input: process.stdin})
    rl.on('line', (line) => {
        source += line + "\n";
    });

    return new Promise<string>((resolve) => {
        rl.on('close', () => {
            resolve(source)
        })
    });
}

const loadInlineScript = (source, argSuiteParams) => {
    const scriptFn = new Function('step', 'suiteParams', 'console', source);
    journey('inline', async () => {
        console.log("MAKING STEPS")
        scriptFn.apply(null, [step, argSuiteParams, console])
    })
    console.log("r STEPS")
}