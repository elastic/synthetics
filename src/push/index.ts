/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */
import { readFile, writeFile } from 'fs/promises';
import { prompt } from 'enquirer';
import { bold, grey } from 'kleur/colors';
import {
  getLocalMonitors,
  buildMonitorSchema,
  diffMonitors as diffMonitorHashIDs,
  MonitorSchema,
} from './monitor';
import { ALLOWED_SCHEDULES, Monitor } from '../dsl/monitor';
import {
  progress,
  liveProgress,
  write,
  error,
  warn,
  indent,
  done,
  getMonitorManagementURL,
  THROTTLING_WARNING_MSG,
} from '../helpers';
import type { PushOptions, ProjectSettings } from '../common_types';
import { findSyntheticsConfig, readConfig } from '../config';
import {
  bulkDeleteMonitors,
  bulkGetMonitors,
  bulkPutMonitors,
  createMonitorsLegacy,
  CHUNK_SIZE,
} from './kibana_api';
import {
  getChunks,
  isBulkAPISupported,
  isLightweightMonitorSupported,
  logDiff,
} from './utils';

export async function push(monitors: Monitor[], options: PushOptions) {
  const duplicates = trackDuplicates(monitors);
  if (duplicates.size > 0) {
    throw error(formatDuplicateError(duplicates));
  }
  progress(
    `Pushing monitors for '${options.id}' project in kibana '${options.space}' space`
  );

  /**
   * Legacy API for kibana which does not support bulk operations
   */
  if (!isBulkAPISupported(options.kibanaVersion)) {
    return await pushLegacy(monitors, options);
  }

  const { monitors: remote } = await bulkGetMonitors(options);

  progress(`bundling ${monitors.length} monitors`);
  const schemas = await buildMonitorSchema(monitors, true);
  const local = getLocalMonitors(schemas);

  const { newIDs, changedIDs, removedIDs, unchangedIDs } = diffMonitorHashIDs(
    local,
    remote
  );
  logDiff(newIDs, changedIDs, removedIDs, unchangedIDs);

  const updatedMonitors = new Set<string>([...changedIDs, ...newIDs]);
  if (updatedMonitors.size > 0) {
    const chunks = getChunks(schemas, CHUNK_SIZE);
    for (const chunk of chunks) {
      await liveProgress(
        bulkPutMonitors(options, chunk),
        `creating or updating ${chunk.length} monitors`
      );
    }
  }

  if (removedIDs.size > 0) {
    if (updatedMonitors.size === 0 && unchangedIDs.size === 0) {
      await promptConfirmDeleteAll(options);
    }
    const chunks = getChunks(Array.from(removedIDs), CHUNK_SIZE);
    for (const chunk of chunks) {
      await liveProgress(
        bulkDeleteMonitors(options, chunk),
        `deleting ${chunk.length} monitors`
      );
    }
  }

  done(`Pushed: ${grey(getMonitorManagementURL(options.url))}`);
}

async function promptConfirmDeleteAll(options: PushOptions) {
  write('');
  const { deleteAll } = await prompt<{ deleteAll: boolean }>({
    type: 'confirm',
    skip() {
      if (options.yes) {
        this.initial = process.env.TEST_OVERRIDE ?? true;
        return true;
      }
      return false;
    },
    name: 'deleteAll',
    message: `Pushing without any monitors will delete all monitors associated with the project.\n Do you want to continue?`,
    initial: false,
  });
  if (!deleteAll) {
    throw warn('Push command Aborted');
  }
}

function trackDuplicates(monitors: Monitor[]) {
  const monitorMap: Map<string, Monitor> = new Map();
  const duplicates: Set<Monitor> = new Set();
  for (const monitor of monitors) {
    const id = monitor.config.id;
    if (monitorMap.has(id)) {
      duplicates.add(monitorMap.get(id));
      duplicates.add(monitor);
    }
    monitorMap.set(id, monitor);
  }
  return duplicates;
}

export function formatDuplicateError(monitors: Set<Monitor>) {
  let outer = bold(`Aborted: Duplicate monitors found\n`);

  let inner = '';
  for (const monitor of monitors) {
    const { config, source } = monitor;
    inner += `* ${config.id} - ${source.file}:${source.line}:${source.column}\n`;
  }
  outer += indent(inner);
  return outer;
}

const INSTALLATION_HELP = `Run 'npx @elastic/synthetics init' to create project with default settings.`;

export async function loadSettings(configPath, ignoreMissing = false) {
  try {
    const config = await readConfig(
      process.env['NODE_ENV'] || 'development',
      configPath
    );
    // Missing config file, fake throw to capture as missing file
    if (Object.keys(config).length === 0) {
      throw '';
    }
    return config.project || ({} as ProjectSettings);
  } catch (e) {
    if (!ignoreMissing) {
      throw error(`Aborted (missing synthetics config file), Project not set up correctly.

${INSTALLATION_HELP}`);
    }
  }
}

export async function validatePush(
  opts: PushOptions,
  settings: ProjectSettings
) {
  validateSettings(opts);
  await catchIncorrectSettings(settings, opts);
}

export function validateSettings(opts: PushOptions) {
  const INVALID = 'Aborted. Invalid synthetics project settings.';

  let reason = '';
  if (!opts.id) {
    reason = `Set project id via
  - CLI '--id <id>'
  - Config file 'project.id' field`;
  } else if (!opts.locations && !opts.privateLocations) {
    reason = `Set default location for all monitors via
  - CLI '--locations <values...> or --privateLocations <values...>'
  - Config file 'monitors.locations' | 'monitors.privateLocations' field`;
  } else if (!opts.schedule) {
    reason = `Set default schedule in minutes for all monitors via
  - CLI '--schedule <mins>'
  - Config file 'monitors.schedule' field`;
  } else if (opts.schedule && !ALLOWED_SCHEDULES.includes(opts.schedule)) {
    reason = `Set default schedule(${
      opts.schedule
    }) to one of the allowed values - ${ALLOWED_SCHEDULES.join(',')}`;
  }

  if (!reason) return;

  throw error(`${INVALID}

${reason}

${INSTALLATION_HELP}`);
}

async function overrideSettings(
  configPath,
  oldValue: string,
  newValue: string
) {
  const cwd = process.cwd();
  configPath = configPath ?? (await findSyntheticsConfig(cwd, cwd));
  if (!configPath) {
    throw warn(`Unable to find synthetics config file: ${configPath}`);
  }
  const config = await readFile(configPath, 'utf-8');
  const updatedConfig = config.replace(
    `id: '${oldValue}'`,
    `id: '${newValue}'`
  );
  await writeFile(configPath, updatedConfig, 'utf-8');
}

export async function catchIncorrectSettings(
  settings: ProjectSettings,
  options: PushOptions
) {
  let override = !settings.id;
  if (settings.id && settings.id !== options.id) {
    // Add an extra line to make it easier to read the prompt
    write('');
    ({ override } = await prompt<{ override: boolean }>({
      type: 'confirm',
      name: 'override',
      skip() {
        if (options.yes) {
          this.initial = process.env.TEST_OVERRIDE ?? true;
          return true;
        }
        return false;
      },
      message: `Monitors were pushed under the '${settings.id}' project. Are you sure you want to push them under the new '${options.id}' (note that this will duplicate the monitors, the old ones being orphaned)`,
      initial: false,
    }));
    if (!override) {
      throw warn('Push command Aborted');
    }
  }
  if (override) {
    await overrideSettings(options.config, settings.id, options.id);
  }
}

export async function pushLegacy(monitors: Monitor[], options: PushOptions) {
  if (isLightweightMonitorSupported(monitors, options)) {
    throw error(
      `Aborted: Lightweight monitors are not supported in ${options.kibanaVersion}. Please upgrade to 8.5.0 or above.`
    );
  }

  let schemas: MonitorSchema[] = [];
  if (monitors.length > 0) {
    progress(`bundling ${monitors.length} monitors`);
    schemas = await buildMonitorSchema(monitors, false);
    const chunks = getChunks(schemas, 10);
    for (const chunk of chunks) {
      await liveProgress(
        createMonitorsLegacy({ schemas: chunk, keepStale: true, options }),
        `creating or updating ${chunk.length} monitors`
      );
    }
  } else {
    await promptConfirmDeleteAll(options);
  }
  await liveProgress(
    createMonitorsLegacy({ schemas, keepStale: false, options }),
    `deleting all stale monitors`
  );

  done(`Pushed: ${grey(getMonitorManagementURL(options.url))}`);
}

// prints warning if any of the monitors has throttling settings enabled during push
export function warnIfThrottled(monitors: Monitor[]) {
  const throttled = monitors.some(monitor => monitor.config.throttling != null);
  if (throttled) {
    warn(THROTTLING_WARNING_MSG);
  }
}
