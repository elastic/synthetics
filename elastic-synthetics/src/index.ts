require('source-map-support').install();

export * from './dsl';
export * from './runner';

import { run } from './runner' ;

if (require.main === module)  {
    run()
}
