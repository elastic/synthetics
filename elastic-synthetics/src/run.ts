#!/usr/bin/env node

require('source-map-support').install();
import { runJourneys } from './runner'
import { Command } from 'commander';
import { readFileSync } from 'fs';
import { journey, step, state } from './dsl';
import { createInterface as createReadlineInterface } from 'readline'

export const run = async (aroundHook: (runSteps: (suiteParams: any) => void, environment: string, suiteParams: any) => void = (r) => { r({}) }) => {
    const program = new Command;
    let argSuiteParams = {};
    program
        .command("run [file]")
        .option('-s, --suite-params <jsonstring>')
        .option('-e, --environment <envname>', 'e.g. production', 'development')
        .option('-j, --json', 'output newline delimited JSON')
        .option('--stdin', 'read script file input from stdin')
        .description("Run the tests")
        .action(async (file, options) => {
            if (options.suiteParams) {
                const parsedSuiteParams = JSON.parse(options.suiteParams);
                argSuiteParams = {...argSuiteParams, ...parsedSuiteParams};
            }

            const singleMode = options.stdin || file;
            if (singleMode) {
                state.reset()
                const source = options.stdin ? await readStdin() : readFileSync(file, {encoding: 'utf8'});
                console.log("Running single script...", source)
                loadInlineScript(source, argSuiteParams);
            }

            console.log("Running...")

            aroundHook((suiteParams) => runJourneys(suiteParams), options.environment, argSuiteParams)
        })

    program.parse()
}

const readStdin = async (): Promise<string> => {
    let source = "";
    const rl = createReadlineInterface({input: process.stdin})
    rl.on('line', (line) => {
        source += line + "\n";
    });

    return new Promise<string>((resolve) => {
        rl.on('close', () => {
            resolve(source)
        })
    });
}

const loadInlineScript = (source, argSuiteParams) => {
    const scriptFn = new Function('step', 'suiteParams', 'console', source);
    journey('inline', async () => {
        console.log("MAKING STEPS")
        scriptFn.apply(null, [step, argSuiteParams, console])
    })
    console.log("r STEPS")
}

run()