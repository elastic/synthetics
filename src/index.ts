/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { runner } from './core';
import { RunOptions } from './core/runner';
import { setLogger } from './core/logger';
import { parseArgs } from './parse_args';

/* eslint-disable @typescript-eslint/no-var-requires */
require('source-map-support').install();

export async function run(options: RunOptions) {
  const cliArgs = parseArgs();
  /**
   * Use the NODE_ENV variable to control the environment if its not explicity
   * passed from either CLI or through the API
   */
  options.environment = options.environment || process.env['NODE_ENV'];
  /**
   * set up logger with appropriate file descriptor
   * to capture all the DEBUG logs when running from heartbeat
   */
  const outfd = options.outfd ?? cliArgs.outfd;
  setLogger(outfd);

  try {
    return await runner.run({
      ...options,
      headless: options.headless ?? cliArgs.headless,
      screenshots: options.screenshots ?? cliArgs.screenshots,
      dryRun: options.dryRun ?? cliArgs.dryRun,
      journeyName: options.journeyName ?? cliArgs.journeyName,
      network: options.network ?? cliArgs.network,
      pauseOnError: options.pauseOnError ?? cliArgs.pauseOnError,
      reporter: cliArgs.json && !options.reporter ? 'json' : options.reporter,
      outfd,
    });
  } catch (e) {
    console.error('Failed to run the test', e);
    process.exit(1);
  }
}

export { beforeAll, afterAll, journey, step, before, after } from './core';
