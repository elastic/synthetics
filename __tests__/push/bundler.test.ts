/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { createReadStream } from 'fs';
import { writeFile, unlink, mkdir, rm } from 'fs/promises';
import unzipper from 'unzipper';
import { join } from 'path';
import { generateTempPath } from '../../src/helpers';
import { Bundler } from '../../src/push/bundler';

async function validateZip(content) {
  const decoded = Buffer.from(content, 'base64');
  const pathToZip = generateTempPath();
  await writeFile(pathToZip, decoded);

  const files = [];
  const entries = createReadStream(pathToZip).pipe(
    unzipper.Parse({ forceStream: true })
  );
  for await (const entry of entries) {
    files.push(entry.path);
  }

  expect(files).toEqual(['__tests__/push/test-bundler/bundle.journey.ts']);
  await unlink(pathToZip);
}

describe('Bundler', () => {
  const PROJECT_DIR = join(__dirname, 'test-bundler');
  const journeyFile = join(PROJECT_DIR, 'bundle.journey.ts');
  const bundler = new Bundler();

  beforeAll(async () => {
    await mkdir(PROJECT_DIR, { recursive: true });
    await writeFile(
      journeyFile,
      `import {journey, step, monitor} from '@elastic/synthetics';
journey('journey 1', () => {
  monitor.use({ id: 'duplicate id' })
  step("step1", () => {})
});`
    );
  });

  afterAll(async () => {
    await rm(PROJECT_DIR, { recursive: true });
  });

  it('build journey', async () => {
    const content = await bundler.build(journeyFile, generateTempPath());
    await validateZip(content);
  });

  it('bundle should be idempotent', async () => {
    const content1 = await bundler.build(journeyFile, generateTempPath());
    const content2 = await bundler.build(journeyFile, generateTempPath());
    expect(content1).toEqual(content2);
  });

  it('throw errors on incorrect path', async () => {
    try {
      await bundler.build(join(PROJECT_DIR, 'blah.ts'), generateTempPath());
    } catch (e) {
      expect(e.message).toContain('ENOENT');
    }
  });
});
