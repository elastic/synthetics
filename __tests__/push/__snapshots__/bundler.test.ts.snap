// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundler build journey 1`] = `
"nullvar __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === \\"object\\" || typeof from === \\"function\\") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, \\"default\\", { value: mod, enumerable: true }) : target, mod));

// node_modules/is-positive/index.js
var require_is_positive = __commonJS({
  \\"node_modules/is-positive/index.js\\"(exports, module) {
    \\"use strict\\";
    module.exports = function(n) {
      return toString.call(n) === \\"[object Number]\\" && n > 0;
    };
  }
});

// asset:/Users/lucasfcosta/Repositories/synthetics/__tests__/push/test-bundler/bundle.journey.ts
var import_is_positive = __toESM(require_is_positive());
import { journey, step, monitor } from \\"@elastic/synthetics\\";
journey(\\"journey 1\\", () => {
  monitor.use({ id: \\"duplicate id\\" });
  step(\\"step1\\", () => {
    (0, import_is_positive.default)(-1);
  });
});
"
`;
